src/test/clojure/clojure/tools/gitlibs/impl_test.cljc

```diff
@@ -32,3 +39,197 @@
+        (#'impl/write-ts lockfile)
+        (Thread/Sleep 1000)
+        (#'impl/write-ts lockfile)
+        
+        ;; Delete lock while git dir exists
+        (#'impl/release-lock lockfile)
+        
+        ;; Wait for process to complete
+        (let [result (wait-for-process proc)]
+          ;; Should return successfully
+          (is (= 0 (:exit result)))
+          (is (str/includes? (:stdout result) (.FullName git-dir-file))))))))
+
+;; Scenario 2: Git dir exists and lockfile doesn't - fast path
+
+(deftest test-scenario-2-fast-path
+  (testing "Git dir exists and lockfile doesn't - fast path"
+    (let [test-url "https://github.com/clojure/spec.alpha.git"
+          git-dir-file (impl/git-dir test-url)
+          lockfile (get-lockfile test-url)
+          config-file-path (get-config-file-path test-url)]
+      
+      ;; Ensure git dir exists
+      (when-not (File/Exists config-file-path)
+        (delete-dir-recursive git-dir-file)
+        (delete-file-if-exists lockfile)
+        (impl/ensure-git-dir test-url))
+      
+      ;; Ensure no lockfile
+      (delete-file-if-exists lockfile)
+      (is (File/Exists config-file-path))
+      (is (not (File/Exists lockfile)))
+      
+      ;; Run in separate process
+      (let [{:keys [proc]} (run-ensure-git-dir-process test-url "/tmp/output2.txt")
+            result (wait-for-process proc)]
+        
+        ;; Should return successfully with the git dir path
+        (is (= 0 (:exit result)))
+        (is (str/includes? (:stdout result) (.FullName git-dir-file)))))))
+
+;; Note: Scenarios 3a and 3b require creating a mock git wrapper which is complex.
+;; For now, I'll create simplified versions that test the core coordination behavior.
+
+(deftest test-scenario-3-simplified-clone-coordination
+  (testing "Lost race: waiter coordinates with cloner process"
+    (let [test-url "https://github.com/clojure/spec.alpha.git"
+          git-dir-file (impl/git-dir test-url)
+          lockfile (get-lockfile test-url)
+          config-file-path (get-config-file-path test-url)]
+      
+      ;; Clean up - simulate starting fresh
+      (delete-dir-recursive git-dir-file)
+      (delete-file-if-exists lockfile)
+      
+      ;; Start the first process (will clone)
+      (let [{:keys [proc]} (run-ensure-git-dir-process test-url "/tmp/output3.txt")]
+        
+        ;; Give it time to acquire lock and start cloning
+        (Thread/Sleep 1000)
+        
+        ;; Verify lock exists
+        (is (File/Exists lockfile))
+        
+        ;; Now our thread tries and should wait
+        (let [result (atom nil)
+              waiter (future
+                       (try
+                         (reset! result (impl/ensure-git-dir test-url))
+                         (catch Exception e
+                           (reset! result {:error (.Message e)}))))]
+          
+          ;; Wait for both to complete
+          (wait-for-process proc)
+          @waiter
```

Comment:
Use the 3 argument deref and give a reasonable timeout that allows the test to succeed but doesn't hang the entire test suite if there's a bug. Ditto in the other tests. Wrap this functionality in a helper like `deref-timely` that throws if timed out.